src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/IndexWriter.java

```diff
@@ -0,0 +1,403 @@
+    /**
+     * Generates the main index of all documented elements, terms defined in some documentation
+     * comments, and summary pages.
+     *
+     * If {@link HtmlOptions#splitIndex()} is true, a separate page is generated for each
+     * initial letter; otherwise, a single page is generated for all items in the index.
+     *
+     * @param configuration the configuration
+     * @throws DocFileIOException if an error occurs while writing the files
+     */
+    public static void generate(HtmlConfiguration configuration) throws DocFileIOException {
+        IndexBuilder mainIndex = configuration.mainIndex;
+        List<Character> firstCharacters = mainIndex.getFirstCharacters();
+        firstCharacters.sort(Comparator.naturalOrder());
+        if (configuration.getOptions().splitIndex()) {
+            ListIterator<Character> iter = firstCharacters.listIterator();
+            while (iter.hasNext()) {
+                Character ch = iter.next();
+                DocPath file = DocPaths.INDEX_FILES.resolve(DocPaths.indexN(iter.nextIndex()));
+                IndexWriter writer = new IndexWriter(configuration, file);
+                writer.generateIndexFile(List.of(ch));
+            }
+        } else {
+            IndexWriter writer = new IndexWriter(configuration, DocPaths.INDEX_ALL);
+            writer.generateIndexFile(firstCharacters);
+        }
+    }
+
+    /**
+     * Creates a writer that can write a page containing some or all of the overall index.
+     *
+     * @param configuration  the current configuration
+     * @param path           the file to be generated
+     */
+    protected IndexWriter(HtmlConfiguration configuration, DocPath path) {
+        super(configuration, path);
+        this.mainIndex = configuration.mainIndex;
+        this.splitIndex = configuration.getOptions().splitIndex();
+    }
+
+    /**
+     * Generates a page containing some or all of the overall index.
+     *
+     * @param firstCharacters the initial characters of the index items to appear on this page
+     * @throws DocFileIOException if an error occurs while writing the page
+     */
+    protected void generateIndexFile(List<Character> firstCharacters) throws DocFileIOException {
+        String title = splitIndex
+                ? resources.getText("doclet.Window_Split_Index", firstCharacters.get(0))
+                : resources.getText("doclet.Window_Single_Index");
+        HtmlTree body = getBody(getWindowTitle(title));
+        Content headerContent = new ContentBuilder();
+        addTop(headerContent);
+        Navigation navBar = new Navigation(null, configuration, PageMode.INDEX, path);
+        navBar.setUserHeader(getUserHeaderFooter(true));
+        headerContent.add(navBar.getContent(Navigation.Position.TOP));
+        Content mainContent = new ContentBuilder();
+        List<Character> allFirstCharacters = new ArrayList<>(mainIndex.getFirstCharacters());
+        allFirstCharacters.sort(Comparator.naturalOrder());
```

Comment:
Given your other comment (listed next in this page) that the sort is actually unnecessary, it may be not so important to pass in the list of all first characters from the `generate` method, but it is easy enough to do so. I will check it out.